Fix DruidNavigator loading multiple tables

kevinhinterlong opened the pull request:
It appears the locking mechanism was forced to time out while loading
multiple tables because the lock for getting all the tables was still active.

Now only the lock for getting all the tables (`loadAllDatasources()`) is used.

kevinhinterlong committed:
DruidNavigator locked up while loading several tables

It appears the locking mechanism was forced to time out while loading
multiple tables

kevinhinterlong committed:
Fix javadoc, revert timeout to 30 seconds

cdeszaq reviewed (dismissed):
A few things here:

1. Changelog entry needed (probably in the `Fixes` section)
2. It would be nice to have a test in place that shows the issue and that this change fixes it.
3. I'm not sure I understand how this is fixing the problem. The main change I see is to move the `get` to the caller. I don't really understand how that solves the problem which, as I understand it, is that the follow-on calls are trying to run while the 1st call isn't finished. Given how the AsyncHttpClient works, multiple calls in flight at once shouldn't be a problem. And there's no lock involved (other than the implicit one in `Future::get`), so I'm a little confused there as well.
4. Don't we still want to have the later calls timeout or at least finish executing before the rest of the application starts up? I think this change would allow the app to start w/o having any of the tables configured...

kevinhinterlong committed:
Fix DruidNavigator loading multiple tables

- This also adds a test to make sure that multiple tables get configured properly.
- There was also an issue with the GenericTableLoader showing the same metrics,
   granularities, etc for all tables

kevinhinterlong committed:
Fix issue where multiple tables load the same dimension name

kevinhinterlong committed:
Fix check style errors

kevinhinterlong committed:
Fix bug where all dimensions displayed for all tables.

Now dimensions show up only in the correct table.
Also made a few fields final.

kevinhinterlong committed:
Fix refactoring which didn't get noticed in a test spec

michael-mclawhorn commented on fili-generic-example/src/main/java/com/yahoo/wiki/webservice/data/config/dimension/GenericDimensionConfigs.java:
```diff
@@ -17,39 +17,46 @@
 public class GenericDimensionConfigs {
-    private final Set<DimensionConfig> dimensionConfigs;
+    private final Map<String, Set<DimensionConfig>> dimensionConfigs;
 
```
It's not clear what the String in this map is.

kevinhinterlong committed:
Clarify variable names

QubitPi commented on fili-generic-example/src/test/groovy/com/yahoo/wiki/webservice/web/endpoints/MultipleDatasourceDruidConfigLoaderSpec.groovy:
```diff
@@ -0,0 +1,80 @@
+            } else if (druidWebService.lastUrl.equals("/datasources/table2/?full")) {
+                return getFullTable(datasources[1], table2_metrics, table2_dimensions)
+            }
+            println "No response for " + druidWebService.lastUrl
```
Maybe you forgot to delete it?

QubitPi commented on fili-generic-example/src/test/groovy/com/yahoo/wiki/webservice/web/endpoints/AutomaticDruidConfigLoaderSpec.groovy:
```diff
@@ -53,7 +52,7 @@ public class AutomaticDruidConfigLoaderSpec extends Specification {
                 return expectedMetricsAndDimensions
             }
-            return "Unexpected URL"
+            println "No response for " + druidWebService.lastUrl
```
Maybe you forgot to delete this?

QubitPi commented on fili-generic-example/src/main/java/com/yahoo/wiki/webservice/data/config/dimension/GenericDimensionConfigs.java:
```diff
@@ -17,39 +17,46 @@
+
+                    dimensionConfigs.put(dataSourceConfiguration.getName(), tableDimensionConfigs);
+                });
+        ;
```
What's this for?

QubitPi commented on fili-generic-example/src/main/java/com/yahoo/wiki/webservice/data/config/dimension/GenericDimensionConfigs.java:
```diff
@@ -58,7 +65,21 @@ public GenericDimensionConfigs(Supplier<List<? extends DataSourceConfiguration>>
+    /**
+     * Get all the dimension configurations associated with this datasource.
+     *
+     * @param dataSourceConfiguration the datasource configuration's dimensions to load
```
Two spaces before description. Capitalize `the`

QubitPi commented on fili-generic-example/src/main/java/com/yahoo/wiki/webservice/data/config/table/GenericTableLoader.java:
```diff
@@ -21,29 +21,23 @@
-import java.util.List;
-import java.util.Set;
+import javax.validation.constraints.NotNull;
+import java.util.*;
```
Avoid start import

QubitPi commented on fili-generic-example/src/test/groovy/com/yahoo/wiki/webservice/web/endpoints/MultipleDatasourceDruidConfigLoaderSpec.groovy:
```diff
@@ -0,0 +1,80 @@
+    ]
+}
+
+"""
```
Align indentation

kevinhinterlong committed:
Address Jiaqi's comments, add changelog, fix readme

- The druid dimension loader is now causing it to fail because the
tables can not be found in the DataSourceMetadataService
- Enabled DruidDimensionLoader

kevinhinterlong committed:
Clarify instructions in the readme

michael-mclawhorn reviewed (approved):
Looks pretty good

cdeszaq reviewed (commented):
Partial review. Didn't get as far as `GenericDimensionConfigs`.

kevinhinterlong committed:
Address Rick's comments and clarify readme for Wikipedia Example

kevinhinterlong commented on fili-generic-example/README.md:
```diff
@@ -90,4 +95,6 @@ Here are some sample queries that you can run to verify your server:
 2. Select the `pom.xml` file at the root of the project
-3. Run `GenericMain` which can be found in `fili-generic-example` (e.g. right click and choose run)
+3. Under `src/main/resources/applicationConfig.properties`, change `bard__non_ui_druid_broker`, 
+`bard__ui_druid_broker`, `bard__druid_coord`, and other properties as needed.
```
Yep these are set correctly for running it locally and they work with the wikipedia walkthrough. 

I added the following note in `README.md` to clarify:

> **NOTE:** if you're running this locally and haven't changed any settings (like the Wikipedia example) you can **skip step 3**.

kevinhinterlong committed:
Add a note in README for last working version

- Also fix list of required settings to top of readme

kevinhinterlong commented:
I added a note in the readme of the version this was last tested with. 
I think this should be good to merge now.

garyluoex reviewed (approved):
Rebase, and Im ok with it
